packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+import type { AddApprovalRequest } from '@metamask/approval-controller';
+import type { ControllerMessenger } from '@metamask/base-controller';
+import type { InfuraNetworkType } from '@metamask/controller-utils';
+import {
+  ApprovalType,
+  BUILT_IN_NETWORKS,
+  isNetworkType,
+} from '@metamask/controller-utils';
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  NetworkClientId,
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
```

Comment:
This runs into a type error because `createAsyncMiddleware` only accepts `JsonRpcRequest<Params>` as the `req` argument, which is incompatible with the `JsonRpcRequest & ExtraProperties` intersection type.

Reply:
Oh my bad defining both properties as optional does make it work.